kv/union_store.go

```diff
@@ -46,15 +46,53 @@ type UnionStore struct {
 
 // NewUnionStore builds a new UnionStore.
 func NewUnionStore(snapshot Snapshot, opts Options) UnionStore {
-	wbuffer := p.Get().(MemBuffer)
-	lazy := p.Get().(MemBuffer)
+	lazy := &lazyMemBuffer{}
 	return UnionStore{
-		WBuffer:            wbuffer,
+		WBuffer:            &lazyMemBuffer{},
 		Snapshot:           NewCacheSnapshot(snapshot, lazy, opts),
 		lazyConditionPairs: lazy,
 	}
 }
 
+type lazyMemBuffer struct {
+	mb MemBuffer
+}
+
+func (lw *lazyMemBuffer) Get(k Key) ([]byte, error) {
+	if lw.mb == nil {
+		return nil, ErrNotExist
+	}
+
+	return lw.mb.Get(k)
+}
+
+func (lw *lazyMemBuffer) Set(key []byte, value []byte) error {
+	if lw.mb == nil {
+		lw.mb = p.Get().(MemBuffer)
```

Comment:
This pr LGTM.
But p is not a good name. Rename it to memdbPool is better. It is not a local variable but with single char name.